File: app/src/main/java/com/cray/software/passwords/helpers/SyncHelper.java
package com.cray.software.passwords.helpers;

import android.content.Context;
import android.os.Environment;
import android.util.Log;

import com.cray.software.passwords.interfaces.Constants;
import com.cray.software.passwords.notes.NoteItem;
import com.cray.software.passwords.passwords.Password;
import com.google.gson.Gson;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.StringWriter;
import java.io.Writer;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

public class SyncHelper {

    private static final String TAG = "SyncHelper";

    private Context mContext;

    public SyncHelper(Context context) {
        this.mContext = context;
    }

    public void exportPasswords() throws JSONException, IOException {
        List<Password> list = DataProvider.getOriginalData(mContext);
        File sdPath = Environment.getExternalStorageDirectory();
        File sdPathDr = new File(sdPath.toString() + "/Pass_backup/" + Constants.DIR_SD);
        if (!sdPathDr.exists()) {
            sdPathDr.mkdirs();
        }
        for (Password password : list) {
            if (isSdPresent()) {
                String exportFileName = password.getUuId() + Constants.FILE_EXTENSION;
                File file = new File(sdPathDr, exportFileName);
                if (file.exists()) {
                    file.delete();
                }
                FileWriter fw = new FileWriter(file);
                fw.write(new Gson().toJson(password));
                fw.close();
            } else Log.i("reminder-info", "Couldn't find external storage!");
        }
    }

    public void exportNotes() throws JSONException, IOException {
        if (!isSdPresent()) return;
        List<NoteItem> list = DataProvider.getOriginalNotes(mContext);
        File sdPath = Environment.getExternalStorageDirectory();
        File sdPathDr = new File(sdPath.toString() + "/Pass_backup/" + Constants.DIR_SD);
        if (!sdPathDr.exists()) {
            sdPathDr.mkdirs();
        }
        for (NoteItem item : list) {
            String exportFileName = item.getKey() + Constants.FILE_EXTENSION_NOTE;
            File file = new File(sdPathDr, exportFileName);
            if (file.exists()) {
                file.delete();
            }
            FileWriter fw = new FileWriter(file);
            String json = new Gson().toJson(item);
            fw.write(json);
            fw.close();
        }
    }

    public void importObjectsFromJson() throws IOException, JSONException {
        if (isSdPresent()) {
            List<String> namesPass = new ArrayList<>();
            for (Password item : DataProvider.getOriginalData(mContext)) {
                namesPass.add(item.getUuId());
            }
            for (NoteItem item : DataProvider.getOriginalNotes(mContext)) {
                namesPass.add(item.getKey());
            }
            File sdPath = Environment.getExternalStorageDirectory();
            File folder = new File(sdPath.toString() + "/Pass_backup/" + Constants.DIR_SD);
            importFromFolder(folder, namesPass);
            folder = new File(sdPath.toString() + "/Pass_backup/" + Constants.DIR_SD_DBX_TMP);
            importFromFolder(folder, namesPass);
            folder = new File(sdPath.toString() + "/Pass_backup/" + Constants.DIR_SD_GDX_TMP);
            importFromFolder(folder, namesPass);
        }
    }

    private void importFromFolder(File folder, List<String> names) throws IOException, JSONException {
        if (!folder.exists()) return;
        File[] files = folder.listFiles();
        if (files == null) return;
        if (files.length > 0) {
            for (File file1 : files) {
                String fileName = file1.getName();
                int pos = fileName.lastIndexOf(".");
                String fileLoc = folder + "/" + fileName;
                String fileNameS = fileName.substring(0, pos);
                if (!names.contains(fileNameS)) {
                    names.add(fileNameS);
                    FileInputStream stream = new FileInputStream(fileLoc);
                    Writer writer = new StringWriter();
                    char[] buffer = new char[1024];
                    try {
                        BufferedReader reader = new BufferedReader(new InputStreamReader(stream, "UTF-8"));
                        int n;
                        while ((n = reader.read(buffer)) != -1) {
                            writer.write(buffer, 0, n);
                        }
                    } finally {
                        stream.close();
                    }
                    String json = writer.toString();
                    if (fileName.endsWith(Constants.FILE_EXTENSION_NOTE)) {
                        importNote(json);
                    } else {
                        importPassword(json);
                    }
                }
            }
        }
    }

    private void importNote(String jsonText) throws JSONException {
        NoteItem item = new Gson().fromJson(jsonText, NoteItem.class);
        DataProvider.saveNote(mContext, item);
    }

    private void importPassword(String jsonText) throws JSONException {
        try {
            DataProvider.savePassword(mContext, new Gson().fromJson(jsonText, Password.class));
        } catch (Exception e) {
            JSONObject jsonObj = new JSONObject(jsonText);
            String title = null;
            if (!jsonObj.isNull(DataBase.COLUMN_TITLE)) {
                title = jsonObj.getString(DataBase.COLUMN_TITLE);
            }
            String login = null;
            if (!jsonObj.isNull(DataBase.COLUMN_LOGIN)) {
                login = jsonObj.getString(DataBase.COLUMN_LOGIN);
            }
            String password = null;
            if (!jsonObj.isNull(DataBase.COLUMN_PASSWORD)) {
                password = jsonObj.getString(DataBase.COLUMN_PASSWORD);
            }
            String url = null;
            if (!jsonObj.isNull(DataBase.COLUMN_URL)) {
                url = jsonObj.getString(DataBase.COLUMN_URL);
            }
            String comment = null;
            if (!jsonObj.isNull(DataBase.COLUMN_COMMENT)) {
                comment = jsonObj.getString(DataBase.COLUMN_COMMENT);
            }
            String date = null;
            if (!jsonObj.isNull(DataBase.COLUMN_DATE)) {
                date = jsonObj.getString(DataBase.COLUMN_DATE);
            }
            int colorPass = 0;
            if (jsonObj.has(DataBase.COLUMN_TECHNICAL)) {
                try {
                    String color = jsonObj.getString(DataBase.COLUMN_TECHNICAL);
                    if (color != null) {
                        colorPass = Integer.parseInt(color);
                    }
                } catch (ClassCastException e1) {
                    colorPass = jsonObj.getInt(DataBase.COLUMN_TECHNICAL);
                }
            }
            String uuID = null;
            if (!jsonObj.isNull(DataBase.COLUMN_PIC_SEL)) {
                uuID = jsonObj.getString(DataBase.COLUMN_PIC_SEL);
            }
            DataProvider.savePassword(mContext, new Password(title, date, login, comment, url, 0, colorPass, password, uuID));
        }
    }

    public static String generateID() {
        return UUID.randomUUID().toString();
    }

    public static boolean isSdPresent() {
        return android.os.Environment.getExternalStorageState().equals(android.os.Environment.MEDIA_MOUNTED);
    }
}

File: app/src/main/java/com/cray/software/passwords/helpers/SyncHelper.kt
package com.cray.software.passwords.helpers

import android.content.Context
import org.json.JSONException
import java.io.File
import java.io.IOException
import java.util.*

class SyncHelper(private val mContext: Context) {

    @Throws(JSONException::class, IOException::class)
    fun exportPasswords() {
//        val list = DataProvider.getOriginalData(mContext)
//        val sdPath = Environment.getExternalStorageDirectory()
//        val sdPathDr = File(sdPath.toString() + "/Pass_backup/" + Constants.DIR_SD)
//        if (!sdPathDr.exists()) {
//            sdPathDr.mkdirs()
//        }
//        for (password in list) {
//            if (isSdPresent) {
//                val exportFileName = password.uuId!! + Constants.FILE_EXTENSION
//                val file = File(sdPathDr, exportFileName)
//                if (file.exists()) {
//                    file.delete()
//                }
//                val fw = FileWriter(file)
//                fw.write(Gson().toJson(password))
//                fw.close()
//            } else
//                Log.i("reminder-info", "Couldn't find external storage!")
//        }
    }

    @Throws(JSONException::class, IOException::class)
    fun exportNotes() {
//        if (!isSdPresent) return
//        val list = DataProvider.getOriginalNotes(mContext)
//        val sdPath = Environment.getExternalStorageDirectory()
//        val sdPathDr = File(sdPath.toString() + "/Pass_backup/" + Constants.DIR_SD)
//        if (!sdPathDr.exists()) {
//            sdPathDr.mkdirs()
//        }
//        for (item in list) {
//            val exportFileName = item.key!! + Constants.FILE_EXTENSION_NOTE
//            val file = File(sdPathDr, exportFileName)
//            if (file.exists()) {
//                file.delete()
//            }
//            val fw = FileWriter(file)
//            val json = Gson().toJson(item)
//            fw.write(json)
//            fw.close()
//        }
    }

    @Throws(IOException::class, JSONException::class)
    fun importObjectsFromJson() {
//        if (isSdPresent) {
//            val namesPass = ArrayList<String>()
//            for (item in DataProvider.getOriginalData(mContext)) {
//                namesPass.add(item.uuId)
//            }
//            for (item in DataProvider.getOriginalNotes(mContext)) {
//                namesPass.add(item.key)
//            }
//            val sdPath = Environment.getExternalStorageDirectory()
//            var folder = File(sdPath.toString() + "/Pass_backup/" + Constants.DIR_SD)
//            importFromFolder(folder, namesPass)
//            folder = File(sdPath.toString() + "/Pass_backup/" + Constants.DIR_SD_DBX_TMP)
//            importFromFolder(folder, namesPass)
//            folder = File(sdPath.toString() + "/Pass_backup/" + Constants.DIR_SD_GDX_TMP)
//            importFromFolder(folder, namesPass)
//        }
    }

    @Throws(IOException::class, JSONException::class)
    private fun importFromFolder(folder: File, names: MutableList<String>) {
//        if (!folder.exists()) return
//        val files = folder.listFiles() ?: return
//        if (files.size > 0) {
//            for (file1 in files) {
//                val fileName = file1.name
//                val pos = fileName.lastIndexOf(".")
//                val fileLoc = "$folder/$fileName"
//                val fileNameS = fileName.substring(0, pos)
//                if (!names.contains(fileNameS)) {
//                    names.add(fileNameS)
//                    val stream = FileInputStream(fileLoc)
//                    val writer = StringWriter()
//                    val buffer = CharArray(1024)
//                    try {
//                        val reader = BufferedReader(InputStreamReader(stream, "UTF-8"))
//                        var n: Int
//                        while ((n = reader.read(buffer)) != -1) {
//                            writer.write(buffer, 0, n)
//                        }
//                    } finally {
//                        stream.close()
//                    }
//                    val json = writer.toString()
//                    if (fileName.endsWith(Constants.FILE_EXTENSION_NOTE)) {
//                        importNote(json)
//                    } else {
//                        importPassword(json)
//                    }
//                }
//            }
//        }
    }

    @Throws(JSONException::class)
    private fun importNote(jsonText: String) {
//        val item = Gson().fromJson(jsonText, NoteItem::class.java)
//        DataProvider.saveNote(mContext, item)
    }

    @Throws(JSONException::class)
    private fun importPassword(jsonText: String) {
//        try {
//            DataProvider.savePassword(mContext, Gson().fromJson(jsonText, Password::class.java))
//        } catch (e: Exception) {
//            val jsonObj = JSONObject(jsonText)
//            var title: String? = null
//            if (!jsonObj.isNull(DataBase.COLUMN_TITLE)) {
//                title = jsonObj.getString(DataBase.COLUMN_TITLE)
//            }
//            var login: String? = null
//            if (!jsonObj.isNull(DataBase.COLUMN_LOGIN)) {
//                login = jsonObj.getString(DataBase.COLUMN_LOGIN)
//            }
//            var password: String? = null
//            if (!jsonObj.isNull(DataBase.COLUMN_PASSWORD)) {
//                password = jsonObj.getString(DataBase.COLUMN_PASSWORD)
//            }
//            var url: String? = null
//            if (!jsonObj.isNull(DataBase.COLUMN_URL)) {
//                url = jsonObj.getString(DataBase.COLUMN_URL)
//            }
//            var comment: String? = null
//            if (!jsonObj.isNull(DataBase.COLUMN_COMMENT)) {
//                comment = jsonObj.getString(DataBase.COLUMN_COMMENT)
//            }
//            var date: String? = null
//            if (!jsonObj.isNull(DataBase.COLUMN_DATE)) {
//                date = jsonObj.getString(DataBase.COLUMN_DATE)
//            }
//            var colorPass = 0
//            if (jsonObj.has(DataBase.COLUMN_TECHNICAL)) {
//                try {
//                    val color = jsonObj.getString(DataBase.COLUMN_TECHNICAL)
//                    if (color != null) {
//                        colorPass = Integer.parseInt(color)
//                    }
//                } catch (e1: ClassCastException) {
//                    colorPass = jsonObj.getInt(DataBase.COLUMN_TECHNICAL)
//                }
//
//            }
//            var uuID: String? = null
//            if (!jsonObj.isNull(DataBase.COLUMN_PIC_SEL)) {
//                uuID = jsonObj.getString(DataBase.COLUMN_PIC_SEL)
//            }
//            DataProvider.savePassword(mContext, Password(title, date, login, comment, url, 0, colorPass, password, uuID))
//        }

    }

    companion object {

        private val TAG = "SyncHelper"

        fun generateID(): String {
            return UUID.randomUUID().toString()
        }

        val isSdPresent: Boolean
            get() = android.os.Environment.getExternalStorageState() == android.os.Environment.MEDIA_MOUNTED
    }
}
